File: leetcode-algorithms/695. Max Area of Island/Solution.cpp
// Given a non-empty 2D array grid of 0's and 1's, an island is a group of 1's (representing land) connected 4-directionally (horizontal or vertical.) You may assume all four edges of the grid are surrounded by water.
//
// Find the maximum area of an island in the given 2D array. (If there is no island, the maximum area is 0.)
//
// Example 1:
// [[0,0,1,0,0,0,0,1,0,0,0,0,0],
//  [0,0,0,0,0,0,0,1,1,1,0,0,0],
//  [0,1,1,0,1,0,0,0,0,0,0,0,0],
//  [0,1,0,0,1,1,0,0,1,0,1,0,0],
//  [0,1,0,0,1,1,0,0,1,1,1,0,0],
//  [0,0,0,0,0,0,0,0,0,0,1,0,0],
//  [0,0,0,0,0,0,0,1,1,1,0,0,0],
//  [0,0,0,0,0,0,0,1,1,0,0,0,0]]
// Given the above grid, return 6. Note the answer is not 11, because the island must be connected 4-directionally.
// Example 2:
// [[0,0,0,0,0,0,0,0]]
// Given the above grid, return 0.
// Note: The length of each dimension in the given grid does not exceed 50.
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        if (grid.empty()) return 0;
        int maxsize = 0;
        vector<vector<bool>> visited(grid.size(), vector<bool>(grid[0].size(), false));
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if ((grid[i][j] == 1) && (!visited[i][j])) {
                    int local = 0;
                    visit(grid, i, j, visited, local);
                    maxsize = max(maxsize, local);
                }
            }
        }
        return maxsize;
    }


    void visit(vector<vector<int>>& grid, int x, int y, vector<vector<bool>>& visited, int& size) {
        int row = grid.size();
        int col = grid[0].size();
        if ((x < 0) || (y < 0) || (x >= row) || (y >= col) || (visited[x][y]) || (grid[x][y] != 1)) return;

        visited[x][y] = true;
        size++;
        vector<vector<int>> dirs={{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

        for (auto dir :dirs) {
            visit(grid, x + dir[0], y + dir[1], visited, size);
        }
    }
};

File: leetcode-algorithms/695. Max Area of Island/Solution.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
    	int row = grid.length;
    	if (row == 0) return 0;
    	int col = grid[0].length;
    	if (col == 0) return 0;

        boolean [][] visited = new boolean[row][col];
        int [] area = new int[1];
        int maxv = 0;
        for (int i = 0; i < row; i++) {
        	for (int j = 0; j < col; j++) {
        		if (!visited[i][j] && (grid[i][j] == 1)) {
        			clearvisited(visited);
        			area[0] = 0;
        			visit(grid, i, j, visited, area);
        			maxv = Math.max(area[0], maxv);
        		}
        	}
        }

        return maxv;
    }

    private void clearvisited(boolean [][] visited) {
    	for (int i = 0; i < visited.length; i++) {
    		for (int j = 0; j < visited[0].length; j++) {
    			visited[i][j] = false;
    		}
    	}
    }

    private void visit(int[][] grid, int i, int j, boolean[][] visited, int[] area) {
    	int row = grid.length;
    	int col = grid[0].length;
    	if ((i < 0) || (i >= row) || (j < 0) || (j >= col) || visited[i][j] || (grid[i][j] != 1)) return;

    	visited[i][j] = true;
    	area[0]++;
    	int[][] dirs = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

    	for (int[] dir : dirs) {
    		visit(grid, i + dir[0], j + dir[1], visited, area);
    	}
    }
}
